### ISSUE TITLE
get_commodity_map removal breaks code

### ISSUE THREAD

**Issue by tbm** [beancount/beancount#570]
Commit  renamed get_commodity_map() to rename get_commodity_directives():

>  To clarify the purpose of the function and to avoid possible
>  sublte breakage in external code, rename the function to
>  get_commodity_directives().

Unfortunately, this leads to very "unsubtle" breakage of real code that calls _map().

I don't think removing a function in a point release is a good idea.  Can we have _map()
for backwards compatibility with a note somewhere that this will go away in the future.

@dnicolodi

**Comment by tbm** [comment]
Specifically, fava_miler 0.2.1 with beancount 2.3.1 or later gives:

```
  File "/home/tbm/.local/lib/python3.7/site-packages/fava_miler/common/favainvestorapi.py", line 16, in get_commodity_map
    return getters.get_commodity_map(self.ledger.entries)
AttributeError: module 'beancount.core.getters' has no attribute 'get_commodity_map'
```

fava_miler 0.2.2 is out already but imho 2.3.x should remain backwards compatible.

**Comment by dnicolodi** [comment]
The reasoning behind the change was lost in the migration to Github. However, the name change is deliberate, as the commit message says: the behavior of ` get_commodity_directives()` is different from the one of `get_commodity_map()`. Unfortunately the behavior of `get_commodity_map()` was broken is subtle ways and fixing it properly required major run time work that was deemed useless in most useful cases. Thus it was decided to simply drop it and replace it with an implementation that does what was most probably expected by most callers. However, we could not be sure that none of the callers in external project relied on the half broken old behavior. Renaming the function was a way to make sure callers did not call the new implementation expecting the old behavior resulting in bugs possibly hard to diagnose.

It was maybe not a good idea to release this change as a point release, but at the time I was not able to find any caller of `get_commodity_map()` outside beancount itself. Unfortunately the release cannot be undone and thus we should deal with the compatibility break. I don't think introducing a stub with the same name but a different behavior is better than the current situation where things break in a obvious way. Furthermore, for projects that rely on the `get_commodity_map()` it is easy to check if it available, but it is much harder to check if the provided `get_commodity_map()` implements the pre-2.3.1 or the post-2.3.1 behavior.

We have already three releases of beancount with the function removed, thus bringing it back does not do any good to projects that used `get_commodity_map()` as they still need to find a way to work with beancount versions 2.3.1 to 2.3.3. Note that releasing a new version of these projects with dependency declarations that avoid these beancount releases is not a better or easier solution than simply upgrading (maybe conditionally) to the new API.

**Comment by blais** [comment]
I rather agree Daniele. In fact, I think this was the reason for the
renaming indeed.
Note that I only made the alias in v2 (not in master).
I could bring back the old function (and mark it as deprecated).

**Comment by dnicolodi** [comment]
I don't think that bringing it back does any good once we released beancount versions with it removed, as I tried to explain in the long comment above. Judging from the commit history, the only external project that relied on the old API adjusted to the new API back in July, but didn't preserve back compatibility and forgot to introduce a requirement for a recent release of beancount. I think there is anything useful that can be done on the beancount side at this point.

As @tbm points out, we should have probably marked the release 2.4 instead than 2.3.1 but it is too late to fix this.
